### PR TITLE
pythran: initial support

### LINKED ISSUE #6079: [COMPILER REQUEST]: Pythran
### Compiler name

Pythran

### Compiler version

0.15.0

### Compiler language

Python

### Compiler homepage

https://pypi.org/project/pythran/

### Motivation

Pythran is a Python to native code used by (at least) scipy to speedup numeric kernels written in Python. It ca n use either GCC or Clang as a backend.

Having it on compiler explorer would help for both performance exploration and bug reporting.

### PR DESCRIPTION
**Pull request by dkm**
Pythran is a python to c++ transpiler.
Our supports both the C++ and the full link to a share library (a python module).

See https://pypi.org/project/pythran/ for more information.

fixes #6079

### CODE REVIEW

**Comment by dkm** [comment]
@serge-sans-paille this is the last step needed before being able to enable pythran on the live site. Can you please have a look and check there's nothing that looks wrong (for the parts you can review of course... license info, naming, etc).

image

image

File: lib/compilers/pythran.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    constructor(info: PreliminaryCompilerInfo, env) {
+        super(info, env);
+        this.cpp_compiler_root = this.compilerProps<string>(`compiler.${this.compiler.id}.cpp_compiler_root`);
+    }
+
+    override getDefaultExecOptions() {
+        const execOptions = super.getDefaultExecOptions();
+        if (this.cpp_compiler_root) {
+            execOptions.env.PATH = path.join(this.cpp_compiler_root, 'bin');
+            execOptions.env.LD_LIBRARY_PATH = path.join(this.cpp_compiler_root, 'lib');
```
Comment (partouf) [inline lib/compilers/pythran.ts:22]: I think for completeness you also need `lib64`

Comment (dkm) [reply lib/compilers/pythran.ts:22]: oh right. I think I had it at some point, but I must have dropped this part by mistake. Thanks!

**Comment by serge-sans-paille** [comment]
I'd appreciate a screenshot with some compiler option, e.g. `--config
backend.annotate=true` ? Just to make sure they are honored.

The licensing part look ok.

**Comment by dkm** [comment]
> I'd appreciate a screenshot with some compiler option, e.g. `--config backend.annotate=true` ? Just to make sure they are honored.
> 
> The licensing part look ok.

Hmmm, nothing stands out. Is it supposed to be very different?

**Comment by dkm** [comment]
Hmmm, how should this be used? In fact, the current code should be working.

The command I have now is:
```
debug: Execution {"args":["-o","/tmp/compiler-explorer-compiler2024127-348816-1hdq6wt.n0rz/output.s","-E","--config","backend.annotate=true","/tmp/compiler-explorer-compiler2
024127-348816-1hdq6wt.n0rz/example.py"],"command":"/opt/compiler-explorer/pythran/pythran-0.15/bin/pythran", ....
```

So this should be working. And I can't find the doc on this particular option.

**Comment by serge-sans-paille** [comment]
yeah, combined with `-E` there should be some extra comments in the generated c++ file that refer to the original python code.

**Comment by dkm** [comment]
image

There's a default filter that removes comments. So it looks like it's working as expected

**Comment by dkm** [comment]
I'll go ahead and merge it.

**Comment by partouf** [comment]
this is now live, thanks!

(I suspect the binary mode is slightly broken, but we can fix some other time)

**Comment by serge-sans-paille** [comment]
extra ! thanks a lot / merci beaucoup :-)

**Comment by dkm** [comment]
> this is now live, thanks!
> 
> (I suspect the binary mode is slightly broken, but we can fix some other time)

What is broken? If you try to "link to binary" a python without any pythran annotation, you get a compiler error. Maybe that's what you're referring to? If so, this is expected IIUC.

**Comment by partouf** [comment]
I did annotate, but the function was nowhere to be found.

**Pull request merged**